src/enumerable.cr

```diff
@@ -1419,4 +1419,223 @@ module Enumerable(T)
+  # ```
+  # (1..3).zip(4..6, 7..9) do |x, y, z|
+  #   puts "#{x} -- #{y} -- #{z}"
+  # end
+  # ```
+  #
+  # The above produces:
+  #
+  # ```text
+  # 1 -- 4 -- 7
+  # 2 -- 5 -- 8
+  # 3 -- 6 -- 9
+  # ```
+  def zip(*others : Indexable | Iterable | Iterator, &block)
+    Enumerable.zip(self, others) do |elems|
+      yield elems
+    end
+  end
+
+  # Returns an `Array` of tuples populated with the elements of `self` and
+  # *others* traversed in tandem.
+  #
+  # Raises an `IndexError` if any of *others* doesn't have as many elements
+  # as `self`. See `zip?` for a version that yields `nil` instead of raising.
+  #
+  # ```
+  # a = [1, 2, 3]
+  # b = ["a", "b", "c"]
+  #
+  # a.zip(b) # => [{1, "a"}, {2, "b"}, {3, "c"}]
+  # ```
+  #
+  # An example with multiple arguments:
+  #
+  # ```
+  # a = [1, 2, 3]
+  # b = (4..6)
+  # c = 8.downto(3)
+  #
+  # a.zip(b, c) # => [{1, 4, 8}, {2, 5, 7}, {3, 6, 6}]
+  # ```
+  def zip(*others : Indexable | Iterable | Iterator)
+    pairs = Array(typeof(zip(*others) { |e| break e }.not_nil!)).new(size)
+    zip(*others) { |e| pairs << e }
+    pairs
+  end
+
+  # Yields elements of `self` and *others* in tandem to the given block.
+  #
+  # All of the elements in `self` will be yielded: if *others* don't have
+  # that many elements they will be returned as `nil`.
+  #
+  # ```
+  # a = [1, 2, 3]
+  # b = ["a", "b"]
+  #
+  # a.zip?(b) { |x, y| puts "#{x.inspect} -- #{y.inspect}" }
+  # ```
+  #
+  # The above produces:
+  #
+  # ```text
+  # 1 -- "a"
+  # 2 -- "b"
+  # 3 -- nil
+  # ```
+  #
+  # An example with multiple arguments:
+  #
+  # ```
+  # (1..3).zip?(4..5, 7..8) do |x, y, z|
+  #   puts "#{x.inspect} -- #{y.inspect} -- #{z.inspect}"
+  # end
+  # ```
+  #
+  # The above produces:
+  #
+  # ```text
+  # 1 -- 4 -- 7
+  # 2 -- 5 -- 8
+  # 3 -- nil -- nil
+  # ```
+  def zip?(*others : Indexable | Iterable | Iterator)
+    Enumerable.zip?(self, others) do |elems|
+      yield elems
+    end
+  end
+
+  # Returns an `Array` of tuples populated with the elements of `self` and
+  # *others* traversed in tandem.
+  #
+  # All elements in `self` are returned in the Array. If matching elements
+  # in *others* are missing (because they don't have that many elements)
+  # `nil` is returned inside that tuple index.
+  #
+  # ```
+  # a = [1, 2, 3]
+  # b = ["a", "b"]
+  #
+  # a.zip?(b) # => [{1, "a"}, {2, "b"}, {3, nil}]
+  # ```
+  #
+  # An example with multiple arguments:
+  #
+  # ```
+  # a = [1, 2, 3]
+  # b = (4..5)
+  # c = 8.downto(7)
+  #
+  # a.zip?(b, c) # => [{1, 4, 8}, {2, 5, 7}, {3, nil, nil}]
+  # ```
+  def zip?(*others : Indexable | Iterable | Iterator)
+    pairs = Array(typeof(zip?(*others) { |e| break e }.not_nil!)).new(size)
+    zip?(*others) { |e| pairs << e }
+    pairs
+  end
+
+  # :nodoc:
+  def self.zip(main, others : U, &block) forall U
+    {% begin %}
+      # Try to see if we need to create iterators (or treat as iterators)
+      # for every element in `others`.
```

Comment:
```suggestion
      # for every element in *others*.
```

Reply:
These aren't public comments. I usually use backticks to refer to code variables and such.